Main.hs

```diff
@@ -457,36 +454,53 @@ fieldArgInfo f = do
     badness = genericDocError =<< text "Not a record field:" <+> prettyTCM f
 
 
-compileRecord :: Definition -> TCM [Hs.Decl ()]
-compileRecord def = do
-  let r = hsName $ prettyShow $ qnameName $ defName def
-  TelV tel _ <- telViewUpTo (recPars (theDef def)) (defType def)
+compileRecord :: RecordTarget -> Definition -> TCM [Hs.Decl ()]
+compileRecord target def = do
+  TelV tel _ <- telViewUpTo recPars (defType def)
   hd <- addContext tel $ do
     let params = teleArgs tel :: [Arg Term]
     pars <- mapM (showTCM . unArg) $ filter visible params
     return $ foldl (\ h p -> Hs.DHApp () h (Hs.UnkindedVar () $ hsName p))
-                   (Hs.DHead () r)
+                   (Hs.DHead () (hsName rName))
                    pars
-  let help :: Int -> [QName] -> Telescope -> TCM [Hs.ClassDecl ()]
-      help i ns tel = case (ns,splitTelescopeAt i tel) of
+  case target of
+    ToClass -> do
+      classDecls <- compileRecFields classDecl recPars (unDom <$> recFields) recTel
+      return [Hs.ClassDecl () Nothing hd [] (Just classDecls)]
+
+    ToRecord | recNamedCon -> do
+      fieldDecls <- compileRecFields fieldDecl recPars (unDom <$> recFields) recTel
+      let conDecl = Hs.QualConDecl () Nothing Nothing $ Hs.RecDecl () cName fieldDecls
+      return [Hs.DataDecl () (Hs.DataType ()) Nothing hd [conDecl] []]
+
+             | otherwise   -> genericError $
+      "Not supported: " ++ rName ++ " is missing a named constructor."
```

Comment:
You need to `setCurrentRange` here to get a source location on the error, but I can fix that post-merge.